### ISSUE TITLE
Support for being able to add genotypes to a BCF record

### ISSUE THREAD

**Issue by hjeremyli** [rust-bio/rust-htslib#224]
As far as I understand, there is currently no way to update genotypes in a bcf `Record`.

Related:
https://github.com/rust-bio/rust-htslib/blob/ed217c2fe460794d249e58a85db0b0c8a2929e55/src/bcf/record.rs#L360

Since the `GT` field has a `String` format (as specified in the header) I tried to work around this by using `Record.push_format_string()` like so:

```rust
fn main() {
    let mut header = rust_htslib::bcf::header::Header::new();
     header.push_record(format!("##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">").as_bytes());
    header.push_sample("samplename".as_bytes());
    // ... add some other stuff to the header...

    let mut writer = rust_htslib::bcf::Writer::from_path("out.vcf.gz", &header, false, rust_htslib::bcf::Format::VCF).unwrap();
    let mut rec = writer.empty_record();    

    // ... populate chr, pos, id, etc. .... 

    // attempt to add genotype field using fixed string 
    let gt_test = vec!["0/0".as_bytes()];
    rec.push_format_string("GT".as_bytes(), &gt_test).unwrap();

    writer.write(&rec).unwrap();
}
```

However, this yields the following error:

```[E::bcf_format_gt] Unexpected type 7
```

This above works if I replace the tag argument in `push_format_string()` with something other than `"GT"`, which seem to suggest that attempting to edit the GT field using this method is treated differently than for other methods. 

It seems I'm likely missing something here, but is there a straightforward way of updating genotypes for a given record as I attempted to do above with the current release? Something I imagine that would provide similar functionality to the `bcf_update_genotypes` macro in `htslib`.

**Comment by MaltheSR** [comment]
I needed to do the same thing and played around with this for a bit. In case anyone else stumbles on this, the GT FORMAT field is apparently special in that it is marked as a string in the header, but it is actually encoded as an integer. (I found [this](https://github.com/samtools/htslib/issues/54) discussion helpful on that point.) 

Therefore, corresponding to the example above, I've been able to do e.g.:

```
let gt_test: [i32; 2] = [2, 2];
record.push_format_integer("GT".as_bytes(), &gt_test).unwrap();
```

Here, `2` is the encoding for an unphased reference. I haven't found the actual specification of the encoding, but from playing around, I find that `1` is missing, `2` is unphased reference, `3` is phased reference, `4` is unphased alternative, and `5` is phased alternative.

Perhaps documenting this more clearly might be helpful? I'd be happy to help on that point, if needed, but I'm still getting used to Rust and not exactly sure where it should go?

**Comment by brainstorm** [comment]
Yes please @MaltheSR, your help documenting this tricky bit would be great indeed!

Cloning this repo and running [`cargo doc`](https://doc.rust-lang.org/rustdoc/index.html) should give you a glimpse on how to do that. Also feel free to jump into the [rust-bio/#rust-htslib](https://discord.gg/rssQABT) discord server if you have further doubts and questions ;)

**Comment by dlaehnemann** [comment]
@MaltheSR The documentation for the encoding of GT in BCF is in the separate BCF specs over here (exact lines with the description):

https://github.com/samtools/hts-specs/blob/81d95f9e97e8b67d0d635c3015e156415c3aaf6b/BCFv2_qref.tex#L49-L54

Or if you prefer a more readable version, check out the PDF (it's just one page, so quick to find the description):
https://github.com/samtools/hts-specs/blob/master/CRAMv2.1.pdf

And from your description above I think you have come across the VCF spec for GT fields, but for a complete reference, here's the most important infos in the VCF spec:
https://github.com/samtools/hts-specs/blob/81d95f9e97e8b67d0d635c3015e156415c3aaf6b/VCFv4.3.tex#L448-L458

**Comment by dlaehnemann** [comment]
And for tipps on how to write documentation in Rust, check out the links over in the Documentation guidelines of `rust-bio`:
https://github.com/rust-bio/rust-bio#documentation-guidelines

These links take you to all the relevant Rust docs with good examples for doc tests and doc strings...

**Comment by MaltheSR** [comment]
Thank you both for the the pointers on the specs as well as Rust docs.

I was trying to write a clear MWE (for doctest examples) for how to push genotypes onto a record, which might also serve as a template for easily writing examples for the other `push_format_*` methods. Problem is that it's quite hard to keep an MWE _minimal_. To illustrate, I believe something like this is required to create and modify a record:

```
use rust_htslib::bcf::{Format, Writer};
use rust_htslib::bcf::header::Header;
use rust_htslib::bcf::record::GenotypeAllele;

fn main() {
    // Create minimal VCF header with a single contig and a single sample
    let mut header = Header::new();
    let header_contig_line = r#"##contig=<ID=1,length=10>"#;
    header.push_record(header_contig_line.as_bytes());
    let header_gt_line = r#"##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">"#;
    header.push_record(header_gt_line.as_bytes());
    header.push_sample("test_sample".as_bytes());

    // Write uncompressed VCF to stdout with above header and create record
    let mut vcf = Writer::from_stdout(&header, true, Format::VCF).unwrap();
    let mut record = vcf.empty_record();

    // Helper to encode alleles with/without phasing. Use -1 for missing.
    // For more information, see https://git.io/JUaXO for BCFv2 specs.
    fn encode_allele(allele: GenotypeAllele) -> i32 {
        let (allele, phased) = match allele {
            GenotypeAllele::UnphasedMissing => (-1, 0),
            GenotypeAllele::PhasedMissing => (-1, 1),
            GenotypeAllele::Unphased(a) => (a, 0),
            GenotypeAllele::Phased(a) => (a, 1),
        };
        allele + 1 << 1 | phased
    }

    // Set record genotype to 0|1; note first allele is always unphased
    let first_allele = encode_allele(GenotypeAllele::Unphased(0));
    let second_allele = encode_allele(GenotypeAllele::Phased(1));
    let genotype: [i32; 2] = [first_allele, second_allele];
    record.push_format_integer("GT".as_bytes(), &genotype).unwrap();

    // Write record to VCF
    vcf.write(&record).unwrap();
}
```

Two questions in this regard:

1. Is it meaningful to have doctest examples that are this long and "unlocalised" to the code that is documented?
2. Should I maybe add an `encode` function to the `GenotypeAllele` enum to mirror the `from_encoded` method that already exists? It would make updating genotypes a bit more readable by hiding the details of thee BCF spec (as in the above).

**Comment by dlaehnemann** [comment]
Thanks a bunch already for looking at this in detail and starting on more thorough docs!

Regarding `2.`: The reciprocal function of `from_encoded` definitely makes sense as a helper function that will increase readability. A version that is more idiomatic to Rust would be to implement the [`From` trait](https://doc.rust-lang.org/std/convert/trait.From.html) in both directions, i.e.:
```
impl From<i32> for GenotypeAllele {}
impl From<GenotypeAllele> for i32 {}
```
For the second, this would be a breaking change for the API, so this would warrant some more pairs of eyes. But for the newly added `From<i32> for GenotypeAllele` you could go straight for the trait implementation.

In a second step, you could further add a helper / wrapper function to directly push a `GT` field from a vector of `GenotypeAllele`s, so something that would do:
```
let genotype_vec = vec![ GenotypeAllele::Unphased(0), GenotypeAllele::Phased(1) ]
record.push_format_GT(&genotype_vec).unwrap();
```
It would need to implicitly call `push_format_integer()` with `"GT".as_bytes()` set and convert the `GenotypeAllele` vector with the newly added function / From trait implementation.

Regarding the doctests:
This is always a tricky balance, we've had this question all the time during the recent Docathons. One thing you can do, is to put less localised tests that demonstrate the use of several functions at the top of a (sub-)module as module-level documentation using `//!` commenting. This way, on docs.rs they get rendered at the top of a page before the list of all the functions and their respective documentation. And to demonstrate how individual functions work, you can hide setup code that would only distract from the usage docs with `/// # some_necessary_function_call_to_hide()`, see e.g.:
https://rust-lang.github.io/api-guidelines/documentation.html#examples-use--not-try-not-unwrap-c-question-mark

It's always a tricky balance to strike. But if in doubt, simply implement it the way you think best and we'll check it out during review. It's always possible to move these tests around. And as we already see, they are great at pointing to missing implementations and other stuff! :D

**Comment by brainstorm** [comment]
I'm not sure this issue is fully addressed on PR #277 (docs?), or is it @MaltheSR ?

**Comment by MaltheSR** [comment]
The functionality itself should be good to go via the `push_genotypes` method. I'll improve the docs for it when I have the time.

Perhaps one outstanding issue is whether to replace the current
```
pub fn from_encoded(encoded: i32) -> Self
```
with 
```
impl From<i32> for GenotypeAllele {}
```
as suggested by @dlaehnemann above. The inverse conversion `impl From<GenotypeAllele> for i32` was included in #277 already. I'm happy to do it, but I'll leave it to some of you core developers to decide whether this is a breaking change that you think is warranted.

**Comment by dlaehnemann** [comment]
Maybe add the new trait implementation and mark `from_encoded()` as deprecated with a `since` value a couple of versions in the future:
https://doc.rust-lang.org/edition-guide/rust-2018/the-compiler/an-attribute-for-deprecation.html

Optimally, the respective deprecation message directly gives the info how to change old syntax to new syntax.
